src/connectionInfo/connectionManager.ts

```diff
@@ -0,0 +1,159 @@
+export class ConnectionManager extends Disposable {
+	private _initHttpPort;
+	private _initWSPort;
+	private _initHost: string;
+	private _connections: Map<vscode.Uri | undefined, Connection>; // undefined key means no workspace root
+
+	constructor(private readonly _extensionUri: vscode.Uri) {
+		super();
+
+		this._initHttpPort = SettingUtil.GetConfig(this._extensionUri).portNumber;
+		this._initWSPort = this._initHttpPort + 1;
+		this._initHost = SettingUtil.GetConfig(this._extensionUri).hostIP;
+
+		if (!this._validHost(this._initHost)) {
+			this._showIncorrectHostFormatError(this._initHost);
+			this._initHost = DEFAULT_HOST;
+		} else if (
+			vscode.env.remoteName &&
+			vscode.env.remoteName != '' &&
+			this._initHost != DEFAULT_HOST
+		) {
+			vscode.window.showErrorMessage(
+				localize(
+					'hostCannotConnect',
+					'Cannot use the host "{0}" when using a remote connection. Using default {1}.',
+					this._initHost,
+					DEFAULT_HOST
+				)
+			);
+			this._initHost = DEFAULT_HOST;
+		}
+
+		this._connections = new Map<vscode.Uri, Connection>();
+
+		this._register(
+			vscode.workspace.onDidChangeConfiguration((e) => {
+				if (e.affectsConfiguration(SETTINGS_SECTION_ID)) {
+					this._pendingPort = SettingUtil.GetConfig(
+						this._extensionUri
+					).portNumber;
+					this._pendingHost = SettingUtil.GetConfig(this._extensionUri).hostIP;
+				}
+			})
+		);
+	}
+
+	/**
+	 * @description If setting for the initial port is changed, change the port that servers try first
+	 */
+	private set _pendingPort(port: number) {
+		this._initHttpPort = port;
+		this._initWSPort = port + 1;
+	}
+
+	private set _pendingHost(host: string) {
+		if (this._validHost(host)) {
+			this._initHost = host;
+		} else {
+			this._showIncorrectHostFormatError(host);
+			this._initHost = DEFAULT_HOST;
+		}
+	}
+
+	private _validHost(host: string) {
+		return isIPv4(host);
+	}
+	private readonly _onConnected = this._register(
+		new vscode.EventEmitter<ConnectionInfo>()
+	);
+
+	/**
+	 * Fires when a new connection connects
+	 */
+	public readonly onConnected = this._onConnected.event;
+
+	/**
+	 * get connection by workspaceFolder
+	 * @param workspaceFolder
+	 * @returns connection
+	 */
+	public getConnection(
+		workspaceFolder: vscode.WorkspaceFolder | undefined
+	): Connection | undefined {
+		return this._connections.get(workspaceFolder?.uri);
+	}
+
+	/**
+	 * get a connection using its current port number
+	 * @param port
+	 * @returns connection
+	 */
+	public getConnectionFromPort(port: number): Connection | undefined {
+		return this.connections.find((e) => e && e.httpPort === port);
+	}
+
+	/**
+	 * create a connection via workspaceFolder
+	 * @param workspaceFolder
+	 * @returns connection
+	 */
+	public createAndAddNewConnection(
+		workspaceFolder: vscode.WorkspaceFolder | undefined
+	): Connection {
+		const connection = this._register(
+			new Connection(
+				workspaceFolder,
+				this._initHttpPort,
```

Comment:
Is there anything that checks whether this port is free? Also an issue for a different PR

Reply:
See earlier comment